[WebDriver][WPE] Allow reconnections while keeping the view open

lauromoura opened the pull request:
#### 79271e75c2fec165d6badfa191b6b9a6edeee611
<pre>
[WebDriver][WPE] Allow reconnections while keeping the view open
<a href="https://bugs.webkit.org/show_bug.cgi?id=277341">https://bugs.webkit.org/show_bug.cgi?id=277341</a>

Reviewed by Carlos Garcia Campos.

Currently, when we close a session, all open windows are unconditionally
closed, while the spec[1] says it&apos;s optional. Allowing the window to
remain open might be interesting when we connect to an already running
browser, another scenario covered by the WebDriver spec[2] and supported
by WPEWebKit.

For the record, while the usual practice is to let the WebDriver service
manage the browser lifetime, in some cases we might want to start the
browser in a more controlled way, for example, to enable
profiling/debugging, loading a heavy page beforehand, etc.

To allow deleting a session and keep the windows open, this commit adds
a new `Automation.json` command, &quot;deleteSession&quot;, which makes the
UIProcess terminate the automation session and thus the RemoteInspector
connection between the UIProcess and the WebDriver service. This command
also serves as a fallback if somehow the browser fails to close all
windows.

On the browser-level API, this commit also adds a new signal,
`WebKitAutomationSession::will-close` to notify the browser that the
WebDriver client disconnected, and the browser is free to either kill or
keep the view open for new sessions as required.

[1] <a href="https://w3c.github.io/webdriver/#dfn-close-the-session">https://w3c.github.io/webdriver/#dfn-close-the-session</a> (Step 3.5)
[2] <a href="https://w3c.github.io/webdriver/#new-session">https://w3c.github.io/webdriver/#new-session</a> (Step 8 note)

* Source/WebDriver/Session.cpp:
(WebDriver::Session::close): Only close windows for non-remote sessions
and issue the new &quot;deleteSession&quot; automation command.
* Source/WebDriver/SessionHost.cpp:
(WebDriver::SessionHost::isRemoteBrowser const): Added
* Source/WebDriver/SessionHost.h: Ditto
* Source/WebDriver/glib/SessionHostGlib.cpp:
(WebDriver::SessionHost::connectionDidClose): If we lost connection to a
non-remote (i.e. managed) browser, kill it.
(WebDriver::SessionHost::setTargetList): Check for empty target list.
(WebDriver::SessionHost::isRemoteBrowser const): Added
* Source/WebKit/UIProcess/API/glib/WebKitAutomationSession.cpp:
(webkit_automation_session_class_init): Add new &quot;will-close&quot; signal
* Source/WebKit/UIProcess/API/glib/WebKitWebContext.cpp:
(webkitWebContextWillCloseAutomationSession): Emit the new signal
* Source/WebKit/UIProcess/Automation/Automation.json: Added
  &quot;deleteSession&quot; command.
* Source/WebKit/UIProcess/Automation/WebAutomationSession.cpp:
(WebKit::WebAutomationSession::deleteSession): Implemented by requesting
to terminate the connection between the WebDriver service and the
browser.
* Source/WebKit/UIProcess/Automation/WebAutomationSession.h: Ditto.

Canonical link: <a href="https://commits.webkit.org/282489@main">https://commits.webkit.org/282489@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/c069b4b2ff1611fd4d682ac52438ba0c5ab9e650

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/63317 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/42673 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/15914 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/67338 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/13925 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/50361 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/14205 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/51008 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/9622 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/66386 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/39620 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/54829 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/31690 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/36303 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/12178 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/12797 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/57842 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/12506 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/69034 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/7264 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/12110 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/58318 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/7295 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/54899 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/58548 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/14020 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/6054 "Passed tests") | | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/38494 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/39573 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/40685 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/39316 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

carlosgcampos commented on Source/WebKit/UIProcess/API/glib/WebKitWebContext.cpp:
```diff
@@ -144,6 +144,7 @@ enum {
     INITIALIZE_WEB_PROCESS_EXTENSIONS,
     INITIALIZE_NOTIFICATION_PERMISSIONS,
     AUTOMATION_STARTED,
+    AUTOMATION_FINISHED,
```
Automation started is needed in the context because it's the signal that creates the session, but once we have a session and we passed it to the user, I think finished signal belongs to the session object, not the context.

lauromoura replied:
Changed the new event to `WebKitAutomationSession::will-close`

carlosgcampos commented on Source/WebKit/UIProcess/API/glib/WebKitWebContext.cpp:
```diff
@@ -721,6 +723,32 @@ static void webkit_web_context_class_init(WebKitWebContextClass* webContextClass
+     * Note that it will never be emitted if automation is not enabled in @context,
+     * see webkit_web_context_set_automation_allowed() for more details.
+     *
+     * Since: 2.18
```
This should be 2.46

lauromoura replied:
Changed in the new event declaration

carlosgcampos commented on Source/WebKit/UIProcess/API/glib/WebKitWebContext.cpp:
```diff
@@ -721,6 +723,32 @@ static void webkit_web_context_class_init(WebKitWebContextClass* webContextClass
             WEBKIT_TYPE_AUTOMATION_SESSION);
 
+    /**
+     * WebKitWebContext::automation-finished:
```
As I said I think this should be WebKitAutomationSession::closed for example, or even will-close would be more accurate

lauromoura replied:
Changed the new event to `WebKitAutomationSession::will-close`

carlosgcampos commented on Source/WebDriver/glib/SessionHostGlib.cpp:
```diff
@@ -344,19 +346,13 @@ void SessionHost::didStartAutomationSession(GVariant* parameters)
-        m_target = Target();
-        if (m_connectionID) {
-            if (m_socketConnection)
-                m_socketConnection->close();
```
I'm not sure I understand this change, why don't we need to call close anymore?

lauromoura replied:
Good catch. Looks like my initial version was relying on the connection being closed through other means. I added the `close()` call back, leaving the variable cleanup for the `connectionDidClose()` call.

carlosgcampos commented on Source/WebKit/UIProcess/Automation/WebAutomationSession.cpp:
```diff
@@ -375,6 +375,12 @@ Inspector::Protocol::ErrorStringOr<void> WebAutomationSession::closeBrowsingCont
+{
+    terminate();
+    return { };
+}
```
It seems win platform tried to solve this, but patched WebAutomationSession::closeBrowsingContext instead, maybe they should be sending this new message instead? I don't think we should do that in this patch, though, not even in a follow up, but let them know we are adding this at least.

lauromoura replied:
cc'ing @haruhisa-shin (Author of win bits in #24495 )

haruhisa-shin replied:
Thank you for letting me know.
We too will look into whether we can use this new message.

So far, WebDriver of win port seems to work fine after applying this patch.

carlosgcampos commented on Source/WebKit/UIProcess/Automation/WebAutomationSession.h:
```diff
@@ -172,6 +172,7 @@ class WebAutomationSession final : public API::ObjectImpl<API::Object::Type::Aut
     void getBrowsingContext(const Inspector::Protocol::Automation::BrowsingContextHandle&, Ref<GetBrowsingContextCallback>&&);
     void createBrowsingContext(std::optional<Inspector::Protocol::Automation::BrowsingContextPresentation>&&, Ref<CreateBrowsingContextCallback>&&);
     Inspector::Protocol::ErrorStringOr<void> closeBrowsingContext(const Inspector::Protocol::Automation::BrowsingContextHandle&);
+    Inspector::Protocol::ErrorStringOr<void> deleteSession();
```
If this never fails, I guess it could be void instead?

lauromoura replied:
Tried making it void, but my understanding is that `ErrorStringOr<...>` is a requirement of the inspector protocol for sync messages (The other methods returning void are marked as async in `Automation.json`)

carlosgcampos commented on Source/WebKit/UIProcess/API/glib/WebKitWebContext.cpp:
```diff
@@ -323,6 +323,7 @@ void WebKitAutomationClient::requestAutomationSession(const String& sessionIdent
 
 void webkitWebContextWillCloseAutomationSession(WebKitWebContext* webContext)
 {
+    g_signal_emit_by_name(webContext->priv->automationSession.get(), "will-close", 0, nullptr);
```
```suggestion
    g_signal_emit_by_name(webContext->priv->automationSession.get(), "will-close");
```

lauromoura committed:
[WebDriver][WPE] Allow reconnections while keeping the view open
https://bugs.webkit.org/show_bug.cgi?id=277341

Reviewed by Carlos Garcia Campos.

Currently, when we close a session, all open windows are unconditionally
closed, while the spec[1] says it's optional. Allowing the window to
remain open might be interesting when we connect to an already running
browser, another scenario covered by the WebDriver spec[2] and supported
by WPEWebKit.

For the record, while the usual practice is to let the WebDriver service
manage the browser lifetime, in some cases we might want to start the
browser in a more controlled way, for example, to enable
profiling/debugging, loading a heavy page beforehand, etc.

To allow deleting a session and keep the windows open, this commit adds
a new `Automation.json` command, "deleteSession", which makes the
UIProcess terminate the automation session and thus the RemoteInspector
connection between the UIProcess and the WebDriver service. This command
also serves as a fallback if somehow the browser fails to close all
windows.

On the browser-level API, this commit also adds a new signal,
`WebKitAutomationSession::will-close` to notify the browser that the
WebDriver client disconnected, and the browser is free to either kill or
keep the view open for new sessions as required.

[1] https://w3c.github.io/webdriver/#dfn-close-the-session (Step 3.5)
[2] https://w3c.github.io/webdriver/#new-session (Step 8 note)

* Source/WebDriver/Session.cpp:
(WebDriver::Session::close): Only close windows for non-remote sessions
and issue the new "deleteSession" automation command.
* Source/WebDriver/SessionHost.cpp:
(WebDriver::SessionHost::isRemoteBrowser const): Added
* Source/WebDriver/SessionHost.h: Ditto
* Source/WebDriver/glib/SessionHostGlib.cpp:
(WebDriver::SessionHost::connectionDidClose): If we lost connection to a
non-remote (i.e. managed) browser, kill it.
(WebDriver::SessionHost::setTargetList): Check for empty target list.
(WebDriver::SessionHost::isRemoteBrowser const): Added
* Source/WebKit/UIProcess/API/glib/WebKitAutomationSession.cpp:
(webkit_automation_session_class_init): Add new "will-close" signal
* Source/WebKit/UIProcess/API/glib/WebKitWebContext.cpp:
(webkitWebContextWillCloseAutomationSession): Emit the new signal
* Source/WebKit/UIProcess/Automation/Automation.json: Added
  "deleteSession" command.
* Source/WebKit/UIProcess/Automation/WebAutomationSession.cpp:
(WebKit::WebAutomationSession::deleteSession): Implemented by requesting
to terminate the connection between the WebDriver service and the
browser.
* Source/WebKit/UIProcess/Automation/WebAutomationSession.h: Ditto.

Canonical link: https://commits.webkit.org/282489@main
